### DISCUSSION TITLE
Setting a variable as a result of a function in Template Expression

### DISCUSSION THREAD

**Discussion by gm89uk** [MilesCranmer/PySR#756] [Q&A] [answered]
Congratulations on the release of PySR v1.0.0!

When using template expressions, can the result of a function be stored as a variable and then called upon in another function?

e.g. lets say you have a set function:10a^2-13a+3 = y and need to use PySR to explore _a_.  

is there a correct syntax for: combine="((; f,), (x1, x2,)) -> 10*a**2-13*a+3", where f(x1,x2) = a
or would it be: combine="((; f,), (x1, x2,)) -> 10*f(x1,x2)**2-13*f(x1,x2)+3", #i.e. having to repeat f(x1,x2)?

Thanks again

**Comment by MilesCranmer** [comment, marked as answer]
Yeah! The `((; f,), (x1, x2,)) -> 10 * f(x1, x2)^2 - 13*f(x1, x2) + 3` syntax is just Julia syntax for an anonymous function so you can have any kind of logic in the function you want. The one requirement is that expressions have the same number of arguments each time they are called. So if you do `f(x1)`, you can't then do `f(x1, x2)`. But you can totally do something like `f(x1) - f(x2)`.

For example:

```
combine = """
function my_structure(exprs, data)
    (; f) = exprs
    (x1, x2) = data
    f_result = f(x1, x2)
    
    return 10 * f_result ^ 2 - 13 * f_result + 3
end
"""
```

Writing the function as `((; f,), (x1, x2)) -> ...` is just a shorthand way of unpacking into variables.

But yeah you can have whole multi-line functions there!

**Reply by gm89uk** [reply]
Perfect, thank you for such a quick reply, looking forward to playing around with this!

**Comment by gm89uk** [comment]
Sorry @MilesCranmer a bit stuck, and with it being a new feature there isn't much yet in the forums to go off. 

I'm working in Julia rather than PySR to permit global variables to be used in a custom loss function otherwise
Essentially I have a categorial variable that I know with domain knowledge, has a linear effect on y with just one parameter. 

I am interested in finding the expression that finds y, but without the noise of the parameter from the categorial variable. Although parametric expressions seems ideal, the actual problem is more complicated and template expressions is an otherwise better fit (if I can get it to work). I input the y and category variables as part of X so that I could access them in structure. The categorial variable is basically a vector of Ints, one for each group. 

```julia
structure = TemplateStructure{(:f, :c)}(
    ((; f, c), (x1, x2, x3, x4, x5, x6, y, category)) -> begin
        transformed_x1 = c(x1, x3, x4)
        o = f(transformed_x1, x2, x3, x4, x5, x6)
        residuals = y .- o
        unique_category = unique(category)
        mean_residuals_by_category = Dict(
            group => mean(residuals[category .== group]) for group in unique_category
        )
        o .-= [mean_residuals_by_category[category[i]] for i in eachindex(o)]
        return o
    end
)


model = SRRegressor(
    niterations=1000000,
    binary_operators=[+,-,*,/],
    maxsize=40,
    bumper=true,
    turbo=true,
    populations=18,
    expression_type = TemplateExpression,
    expression_options = (; structure),
    population_size=30,
    parsimony = 0.01,
    batching=true,
)
mach = machine(model, X, y)
fit!(mach)
```
Although I think my approach is completely wrong, as structure is dealing with valid_vectors, I am really at a loss of how to get something like this to work.

**Reply by MilesCranmer** [reply]
I think you are almost there! The next thing would be to unpack the ValidVector with `.x` - which gives you the raw data.

Maybe this would work?

```julia
structure = TemplateStructure{(:f, :c)}(
    ((; f, c), (x1, x2, x3, x4, x5, x6, y, category)) -> begin
        transformed_x1 = c(x1, x3, x4)
        o = f(transformed_x1, x2, x3, x4, x5, x6)
        residuals = y - o  # Changed
        unique_category = unique(category.x)  # Changed
        _mean(_x) = sum(_x) / length(_x)  # Just in case it doesn't exist
        mean_residuals_by_category = Dict(
            group => _mean(residuals.x[category.x .== group])  # Changed
            for group in unique_category
        )
        offsets = [mean_residuals_by_category[category[i]] for i in eachindex(o.x)]  # Changed
        return ValidVector(o.x .- offsets, o.valid && residuals.valid)
    end
)
```

Basically each `x1, x2, x3, ...` is a `ValidVector`. This has a `.x` field which stores the raw data vector, and a `.valid` which stores whether the evaluation encountered any NaNs/Infs. This is because invalid values are very common in evolution, and it's faster to pass them through as bools rather than throw errors and catch them.

The `ValidVector` doubles a convenience wrapper around `Vector` to automatically perform vectorized operations. So you can just write `-` between two `ValidVector` objects, like `x1 - x2` and it will do `x1.x .- x2.x` and `x1.valid && x2.valid`.

When you evaluate an expression like `c(x1, x3, x4)`, the result is another `ValidVector`.

Changes:
- `y .- o` into `y - o`.
- `category` into `category.x`
- `residuals` into `residuals.x` within the `Dict`
- Added the `_mean` function (`mean` doesn't exist in Julia due to the ambiguity between sample/population means)
- Changed the `o .-= offsets` into a `o.x .- offsets`
- I also did an explicit `ValidVector` as a return value just as a demonstration, and so you could include the `.valid` part of the `residual` `ValidVector` (though I think here it's probably not necessary, but just a good habit!)
